pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/LightningKokkosSimulator.cpp

```diff
@@ -0,0 +1,575 @@
+    const std::vector<QubitIdType> &wires) -> ObsIdType {
+    RT_FAIL_IF(wires.size() > this->GetNumQubits(), "Invalid number of wires");
+    RT_FAIL_IF(!isValidQubits(wires), "Invalid given wires");
+
+    auto &&dev_wires = getDeviceWires(wires);
+
+    if (id == ObsId::Hermitian) {
+        return this->obs_manager.createHermitianObs(matrix, dev_wires);
+    }
+
+    return this->obs_manager.createNamedObs(id, dev_wires);
+}
+
+auto LightningKokkosSimulator::TensorObservable(
+    const std::vector<ObsIdType> &obs) -> ObsIdType {
+    return this->obs_manager.createTensorProdObs(obs);
+}
+
+auto LightningKokkosSimulator::HamiltonianObservable(
+    const std::vector<double> &coeffs, const std::vector<ObsIdType> &obs)
+    -> ObsIdType {
+    return this->obs_manager.createHamiltonianObs(coeffs, obs);
+}
+
+auto LightningKokkosSimulator::Expval(ObsIdType obsKey) -> double {
+    RT_FAIL_IF(!this->obs_manager.isValidObservables({obsKey}),
+               "Invalid key for cached observables");
+
+    // update tape caching
+    if (this->tape_recording) {
+        cache_manager.addObservable(obsKey, MeasurementsT::Expval);
+    }
+
+    auto &&obs = this->obs_manager.getObservable(obsKey);
+
+    Pennylane::LightningKokkos::Measures::Measurements<StateVectorT> m{
+        *(this->device_sv)};
+
+    return device_shots ? m.expval(*obs, device_shots, {}) : m.expval(*obs);
+}
+
+auto LightningKokkosSimulator::Var(ObsIdType obsKey) -> double {
+    RT_FAIL_IF(!this->obs_manager.isValidObservables({obsKey}),
+               "Invalid key for cached observables");
+
+    // update tape caching
+    if (this->tape_recording) {
+        this->cache_manager.addObservable(obsKey, MeasurementsT::Var);
+    }
+
+    auto &&obs = this->obs_manager.getObservable(obsKey);
+
+    Pennylane::LightningKokkos::Measures::Measurements<StateVectorT> m{
+        *(this->device_sv)};
+
+    return device_shots ? m.var(*obs, device_shots) : m.var(*obs);
+}
+
+void LightningKokkosSimulator::State(DataView<std::complex<double>, 1> &state) {
+    using UnmanagedComplexHostView =
+        Kokkos::View<Kokkos::complex<double> *, Kokkos::HostSpace,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+    const size_t num_qubits = this->device_sv->getNumQubits();
+    const size_t size = Pennylane::Util::exp2(num_qubits);
+    RT_FAIL_IF(state.size() != size,
+               "Invalid size for the pre-allocated state vector");
+
+    // create a temporary buffer to copy the underlying state-vector to
+    std::vector<std::complex<double>> buffer(size);
+    auto *state_kptr =
+        reinterpret_cast<Kokkos::complex<double> *>(buffer.data());
+
+    // copy data from device to host
+    auto device_data = this->device_sv->getView();
+    Kokkos::deep_copy(UnmanagedComplexHostView(state_kptr, size), device_data);
+
+    // move data to state leveraging MemRefIter
+    std::move(buffer.begin(), buffer.end(), state.begin());
```

Comment:
Feels like this should be handle by some `DeviceToHost` overload.
```suggestion
    const size_t num_qubits = this->device_sv->getNumQubits();
    const size_t size = Pennylane::Util::exp2(num_qubits);
    auto device_data = this->device_sv->DeviceToHost(&(*state.begin()), size);
```

Reply:
Nice!